ops/db-connect.sh

```diff
@@ -1,30 +1,135 @@
+# addition, you MUST be connected to the VPN to access the database.
 #
-# If you get an AlreadyExists error, it's possible the pod didn't shut down properly. You can
-# delete it using kubectl, where ZZ is the SUFFIX.
-# $ kubectl --namespace ZZ delete pod ZZ-psql
+# See usage section below for more details. All arguments are optional.
+#
+
+set -eu
+
+usage() {
+  cat <<EOF
+Usage: $0 [OPTION]...
+Connect to a Terra Data Repo database in an environment.
+
+  --env ENV             Environment to connect to (default: dev)
+  --port PORT           Local port to forward to the SQL proxy (default: 5432)
+  --database DATABASE   Database to connect to, either datarepo or stairway
+                        (default: datarepo)
+  --help                Display this help and exit
+EOF
+  exit 0
+}
+
+error() {
+  echo "ERROR: $1" >&2
+  exit 1
+}
+
+# default values that may be overridden by command line arguments or environment variables
+ENV="${ENV:-dev}"
+PORT="${PORT:-5432}"
+DATABASE="${DATABASE:-datarepo}"
+
+parse_cli_args() {
+  while [ $# -gt 0 ]; do
+    case "$1" in
+      --env)
+        ENV="$2"
+        shift 2
+        ;;
+      --port)
+        PORT="$2"
+        shift 2
+        ;;
+      --database)
+        if [ "$2" != "datarepo" ] && [ "$2" != "stairway" ]; then
+          error "Database must be one of 'datarepo' or 'stairway'"
+        fi
+        DATABASE="$2"
+        shift 2
+        ;;
+      --help)
+        usage
+        ;;
+      *)
+        error "Unknown option: $1. Try --help to see a list of all options."
+        ;;
+    esac
+  done
+}
+
+cleanup() {
+  kill "$PID"
+}
+
+set_vars_from_env() {
+  case "$ENV" in
+    dev)
+      PROJECT="broad-jade-dev"
+      NAMESPACE="dev"
+      SECRET="helm-datarepodb"
+      ;;
+    staging)
+      PROJECT="terra-datarepo-staging"
+      NAMESPACE="terra-staging"
+      SECRET="sql-db"
+      ;;
+    *)
+      error "Unknown environment: $ENV"
+      ;;
+  esac
+}
+
+set_project_config() {
+  gcloud config set project "$PROJECT"
+}
+
+set_cluster_credentials() {
+  CLUSTER_JSON=$(gcloud container clusters list --format="json")
+
+  REGION=$(echo "$CLUSTER_JSON" | jq -r .[0].zone)
+  NAME=$(echo "$CLUSTER_JSON" | jq -r .[0].name)
+
+  gcloud container clusters get-credentials "$NAME" --region="$REGION" --project="$PROJECT"
+}
+
+port_forward_sqlproxy() {
+  POD_JSON=$(kubectl get pods --namespace="$NAMESPACE" --output="json")
+
+  # validate that the namespace is in the list of namespaces
+  if ! echo "$POD_JSON" | jq -r '.items[].metadata.namespace' | grep -q "$NAMESPACE"; then
+    error "Namespace '$NAMESPACE' not found in list of namespaces"
+  fi
 
-: "${DB:?}"
-: "${ENVIRONMENT:?}"
-SUFFIX=${SUFFIX:-$ENVIRONMENT}
+  # select the first pod that has a name that contains "sqlproxy"
+  SQLPROXY_POD=$(kubectl get pods --namespace="$NAMESPACE" --output="json" | jq -r '[ .items[].metadata.name | select(. | contains("sqlproxy")) ].[0]')
```

Comment:
I pulled this feature branch down and tried to run the script, but ran into a compile error that I believe is related to this line:

```
(base) okotsopo@wm111-e35 jade-data-repo % ./ops/db-connect.sh --env dev
Updated property [core/project].
Fetching cluster endpoint and auth data.
kubeconfig entry generated for dev-master.
jq: error: syntax error, unexpected '[', expecting FORMAT or QQSTRING_START (Unix shell quoting issues?) at <top-level>, line 1:
[ .items[].metadata.name | select(. | contains("sqlproxy")) ].[0]                                                              
jq: 1 compile error
```

Reply:
Hrm, I wonder if there's a `jq` version issue. Which version of `jq` are you on? My `jq` version is 1.7.1.

Reply:
Actually, I wonder if this is shell issue, which shell is executing the script? My guess is it's not following POSIX semantics.